crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -1886,9 +1888,25 @@ where
                     },
                 );
 
-                // If any global bindings don't already exist in the global scope, add it.
+                let definition = docstrings::extraction::extract_definition(
+                    ExtractionTarget::Function,
+                    stmt,
+                    self.semantic_model.definition_id,
+                    &self.semantic_model.definitions,
+                );
+                self.semantic_model.push_definition(definition);
+
+                self.semantic_model.push_scope(match &stmt {
+                    Stmt::FunctionDef(stmt) => ScopeKind::Function(stmt),
+                    Stmt::AsyncFunctionDef(stmt) => ScopeKind::AsyncFunction(stmt),
+                    _ => unreachable!("Expected Stmt::FunctionDef | Stmt::AsyncFunctionDef"),
+                });
```

Comment:
I think it could make sense for us to introduce a `AnyFunctionDef` union in `rust_python_ast` that has accessor methods for every field (that both `FunctionDef` and `AsyncFunctionDef` have. It can implement `From`, the `AstNode` trait, and `as/into_function_def()` and `as/into_async_function_def()` methods. 

We found this a very powerful pattern at Rome and I think that would simplify the usage here. 

https://github.com/rome/tools/blob/38104b339da8ff688f469799e1fc3f4a46f3d2ec/crates/rome_js_syntax/src/generated/nodes.rs#L13369-L13404
https://github.com/rome/tools/blob/38104b339da8ff688f469799e1fc3f4a46f3d2ec/crates/rome_js_syntax/src/union_ext.rs#L132-L238

Reply:
What if we just used a single `FunctionDef` node with an `async_` property?

Reply:
That's not CPython compatible ;)

Reply:
Pshhhh